### PR TITLE
sound: remove ctrl msg param

### PR DESCRIPTION
**Pull request by MatiasVara**
### Summary of the PR

This PR is the result of splitting https://github.com/rust-vmm/vhost-device/pull/556. This PR addresses the required changes to remove the `ControlMessage` parameter from the `set_param()` and `release()` functions. In particular in `set_param()`, this PR replaces it with the `VirtioSndPcmSetParams` structure that is parsed in the device thus preventing the audio backend to manipulate descriptors. 

Thanks for reviewing!

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: staging/vhost-device-sound/src/device.rs [hunk 1]
```diff
@@ -322,21 +322,8 @@ impl VhostUserSoundThread {
...
-                                ControlMessage {
-                                    kind: code,
-                                    code: VIRTIO_SND_S_OK,
-                                    desc_chain,
-                                    descriptor: desc_hdr,
-                                    vring: vring.clone(),
-                                },
-                            )
+                            .set_parameters(stream_id, request)
                             .unwrap();
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:326]: We should remove the unwrap() here and set the appropriate response (`VIRTIO_SND_S_BAD_MSG`, `VIRTIO_SND_S_NOT_SUPP`). If the request returns `Err(Error::UnexpectedAudioBackendConfiguration)` the daemon panics. Let's add a match on possible errors and set the respond code accordingly.

File: staging/vhost-device-sound/src/device.rs [hunk 2]
```diff
@@ -351,24 +351,7 @@ impl VhostUserSoundThread {
...
-                                    descriptor: desc_hdr,
-                                    vring: vring.clone(),
-                                },
-                            )
-                            .unwrap();
-
-                        // PcmRelease needs to flush IO messages; the audio backend will reply when
-                        // it drops the ControlMessage.
-                        continue;
+                        audio_backend.write().unwrap().release(stream_id).unwrap();
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:354]: Same as previous commit, match on error to set `resp.code`

File: staging/vhost-device-sound/src/device.rs (hunk 1 above)
Comment (MatiasVara) [reply staging/vhost-device-sound/src/device.rs:326]: Thank you for catching up!

File: staging/vhost-device-sound/src/device.rs (hunk 2 above)
Comment (MatiasVara) [reply staging/vhost-device-sound/src/device.rs]: The only difference here is that I replaced `VIRTIO_SND_S_BAD_MSG` with `VIRTIO_SND_S_IO_ERR` when the state transition is invalid.

Comment (epilys) [reply staging/vhost-device-sound/src/device.rs]: I think yes :thinking: We need to keep the error semantics, i.e. return IO error when we can't read/write from sound backends.

File: staging/vhost-device-sound/src/device.rs [hunk 3]
```diff
@@ -318,25 +318,23 @@ impl VhostUserSoundThread {
...
-                        // PcmSetParams needs check valid formats/rates; the audio backend will
-                        // reply when it drops the ControlMessage.
-                        continue;
+                    } else if let Err(err) = audio_backend
+                        .read()
+                        .unwrap()
+                        .set_parameters(stream_id, request)
+                    {
+                        match err {
+                            Error::StreamWithIdNotFound(_) => {
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:327]: `set_parameters` returns `Error::Stream` as well:

```suggestion
                            Error::Stream(_) | Error::StreamWithIdNotFound(_) => {
```

File: staging/vhost-device-sound/src/device.rs [hunk 4]
```diff
@@ -363,25 +361,17 @@ impl VhostUserSoundThread {
...
-
-                        // PcmRelease needs to flush IO messages; the audio backend will reply when
-                        // it drops the ControlMessage.
-                        continue;
+                    } else if let Err(err) = audio_backend.write().unwrap().release(stream_id) {
+                        match err {
+                            Error::StreamWithIdNotFound(_) => {
+                                resp.code = VIRTIO_SND_S_BAD_MSG.into()
+                            }
+                            Error::Stream(_) => resp.code = VIRTIO_SND_S_BAD_MSG.into(),
```
Comment (epilys) [inline staging/vhost-device-sound/src/device.rs:369]:
```suggestion
                            Error::Stream(_) | Error::StreamWithIdNotFound(_) => {
                                resp.code = VIRTIO_SND_S_BAD_MSG.into()
                            }
```

**Review by epilys — APPROVED** [review]
LGTM, haven't tested yet though, I will try in the morning!

**Review by vireshk — APPROVED** [review]

**Comment by epilys** [comment]
FYI did not get to test it, I just woke up :) no problem of course, I think it looks fine and if there's anything we can fix it afterwards.

**Pull request merged**